File: python-sdk/tests/test_admin_rpcs.py
"""
Tests for the admin RPC wrappers: Session.kill, AsyncSession.kill,
kill_session(), and daemon_health().

Uses the MockDaemon fixture from conftest.py with custom handlers for
session.kill and daemon.health.
"""

from __future__ import annotations

import asyncio
from typing import Any

import loom


def _register_admin_handlers(daemon: Any, *, deep_shims: list[dict] | None = None) -> dict[str, Any]:
    """Wire session.kill + daemon.health handlers into MockDaemon."""
    state: dict[str, Any] = {"kill_calls": [], "health_calls": []}

    def _session_kill(params: dict) -> dict:
        sid = params.get("session_id", "")
        state["kill_calls"].append(sid)
        if sid in daemon.sessions:
            daemon.sessions[sid]["status"] = "killed"
            return daemon.sessions[sid].copy()
        return {"session_id": sid, "status": "killed", "created_at_ms": 0}

    def _daemon_health(params: dict) -> dict:
        state["health_calls"].append(params)
        out = {
            "active_sessions": len(
                [s for s in daemon.sessions.values() if s["status"] == "active"]
            ),
            "shim_breaker_states": [],
            "otel_exporter": "disabled",
            "deep": None,
        }
        if params.get("deep") and deep_shims is not None:
            out["deep"] = deep_shims
        return out

    daemon.register_handler("session.kill", _session_kill)
    daemon.register_handler("daemon.health", _daemon_health)
    return state


def test_session_kill_method_via_handle(daemon):
    state = _register_admin_handlers(daemon)
    with loom.Session.create(
        socket_path=str(daemon.socket_path), token=daemon.token
    ) as s:
        s.kill()
    assert state["kill_calls"], "session.kill RPC was not invoked"
    assert state["kill_calls"][0].startswith("01TEST")


def test_kill_session_free_function(daemon):
    state = _register_admin_handlers(daemon)
    # Create a session first so we have an id to kill.
    with loom.Session.create(
        socket_path=str(daemon.socket_path), token=daemon.token
    ) as s:
        sid = s.session_id
    # Now kill it admin-style (separate connection).
    loom.kill_session(sid, socket_path=str(daemon.socket_path), token=daemon.token)
    # Two kill calls: one from `with` exit (close not kill), one from kill_session
    # Actually close is session.close not session.kill, so only ONE kill call:
    assert len(state["kill_calls"]) == 1
    assert state["kill_calls"][0] == sid


def test_daemon_health_shallow(daemon):
    _register_admin_handlers(daemon)
    result = loom.daemon_health(
        socket_path=str(daemon.socket_path), token=daemon.token
    )
    assert "active_sessions" in result
    assert "shim_breaker_states" in result
    assert "otel_exporter" in result
    assert result["deep"] is None


def test_daemon_health_deep(daemon):
    deep_payload = [
        {
            "shim_id": "chromium-01",
            "daemon_restart_count": 2,
            "daemon_last_restart_at_ms": 1730000000000,
            "shim_uptime_ms": 12345,
            "shim_requests_served": 7,
            "shim_last_request_at_ms": 1730000005000,
            "probe_status": "ok",
        }
    ]
    _register_admin_handlers(daemon, deep_shims=deep_payload)
    result = loom.daemon_health(
        deep=True, socket_path=str(daemon.socket_path), token=daemon.token
    )
    assert result["deep"] is not None
    assert len(result["deep"]) == 1
    assert result["deep"][0]["shim_id"] == "chromium-01"
    assert result["deep"][0]["probe_status"] == "ok"
    assert result["deep"][0]["shim_uptime_ms"] == 12345


def test_async_session_kill(daemon):
    _register_admin_handlers(daemon)

    async def _run() -> None:
        s = await loom.AsyncSession.create(
            socket_path=str(daemon.socket_path), token=daemon.token
        )
        try:
            await s.kill()
        finally:
            await s._transport.close()

    asyncio.run(_run())


def test_kill_session_nonexistent_returns_rpc_error(daemon):
    """Negative path: kill_session on an unknown id should surface a
    typed LoomRPCError (not a connection error or silent success)."""
    import pytest

    # Override handler to error on unknown session ids.
    def _kill_strict(params: dict) -> dict:
        sid = params.get("session_id", "")
        if sid not in daemon.sessions:
            raise ValueError(f"session not found: {sid}")  # MockDaemon maps to internal_error envelope
        daemon.sessions[sid]["status"] = "killed"
        return daemon.sessions[sid].copy()

    daemon.register_handler("session.kill", _kill_strict)
    with pytest.raises(loom.LoomRPCError):
        loom.kill_session(
            "01NONEXISTENT" + "0" * 12,
            socket_path=str(daemon.socket_path),
            token=daemon.token,
        )


def test_daemon_health_error_envelope_surfaces_as_rpc_error(daemon):
    """Negative path: daemon.health returning an error envelope should
    raise LoomRPCError, not silently return a malformed result."""
    import pytest

    def _broken_health(_params: dict) -> dict:
        raise RuntimeError("simulated daemon-side failure")

    daemon.register_handler("daemon.health", _broken_health)
    with pytest.raises(loom.LoomRPCError):
        loom.daemon_health(socket_path=str(daemon.socket_path), token=daemon.token)

File: typescript-sdk/tests/test_admin_rpcs.ts
/**
 * Tests for the admin RPC wrappers: Session.kill, killSession,
 * daemonHealth + AbortSignal integration.
 */
import { test, describe, before, after } from "node:test";
import assert from "node:assert/strict";
import {
  Session,
  LoomTransport,
  LoomAbortError,
  killSession,
  daemonHealth,
} from "../src/index.js";
import { MockDaemon } from "./helpers/mock_daemon.js";

describe("Admin RPCs: session.kill, daemon.health, AbortSignal", () => {
  let daemon: MockDaemon;
  const killCalls: string[] = [];
  const healthCalls: Record<string, unknown>[] = [];

  before(async () => {
    daemon = new MockDaemon();
    daemon.registerHandler("session.kill", (params) => {
      const sid = (params["session_id"] as string) ?? "";
      killCalls.push(sid);
      const s = daemon.sessions.get(sid);
      if (s) {
        s["status"] = "killed";
        return { ...s };
      }
      return { session_id: sid, status: "killed", created_at_ms: 0 };
    });
    daemon.registerHandler("daemon.health", (params) => {
      healthCalls.push(params);
      const out: Record<string, unknown> = {
        active_sessions: Array.from(daemon.sessions.values()).filter(
          (s) => s["status"] === "active",
        ).length,
        shim_breaker_states: [],
        otel_exporter: "disabled",
        deep: null,
      };
      if (params["deep"]) {
        out["deep"] = [
          {
            shim_id: "chromium-test-01",
            daemon_restart_count: 2,
            daemon_last_restart_at_ms: 1_730_000_000_000,
            shim_uptime_ms: 12_345,
            shim_requests_served: 7,
            shim_last_request_at_ms: 1_730_000_005_000,
            probe_status: "ok",
          },
        ];
      }
      return out;
    });
    await daemon.start();
  });

  after(async () => {
    await daemon.stop();
  });

  test("Session.kill() calls session.kill RPC and SIGKILLs the session", async () => {
    const before = killCalls.length;
    const s = await Session.create({
      socketPath: daemon.socketPath,
      token: daemon.token,
    });
    await s.kill();
    assert.strictEqual(killCalls.length, before + 1);
    assert.ok(killCalls[before].startsWith("01TEST"));
  });

  test("killSession() free function targets a sessionId without a handle", async () => {
    // Create a session via one connection.
    const s = await Session.create({
      socketPath: daemon.socketPath,
      token: daemon.token,
    });
    const sid = s.sessionId;
    await s.close();
    // Kill it admin-style via a separate connection.
    const before = killCalls.length;
    await killSession(sid, { socketPath: daemon.socketPath, token: daemon.token });
    assert.strictEqual(killCalls.length, before + 1);
    assert.strictEqual(killCalls[before], sid);
  });

  test("daemonHealth() shallow: returns active_sessions + breaker states + deep:null", async () => {
    const before = healthCalls.length;
    const h = await daemonHealth({
      socketPath: daemon.socketPath,
      token: daemon.token,
    });
    assert.ok(typeof h.activeSessions === "number");
    assert.ok(Array.isArray(h.shimBreakerStates));
    assert.strictEqual(h.deep, null);
    assert.strictEqual(healthCalls.length, before + 1);
    assert.strictEqual(healthCalls[before]["deep"], false);
  });

  test("daemonHealth({deep:true}): typed ShimDeepHealth array with probe_status enum", async () => {
    const h = await daemonHealth({
      deep: true,
      socketPath: daemon.socketPath,
      token: daemon.token,
    });
    assert.ok(h.deep);
    assert.strictEqual(h.deep.length, 1);
    const s = h.deep[0];
    assert.strictEqual(s.shimId, "chromium-test-01");
    assert.strictEqual(s.daemonRestartCount, 2);
    assert.strictEqual(s.shimUptimeMs, 12_345);
    assert.strictEqual(s.probeStatus, "ok");
  });

  test("AbortSignal pre-aborted: rejects synchronously with LoomAbortError", async () => {
    const t = new LoomTransport(daemon.socketPath, daemon.token);
    await t.connect();
    try {
      const controller = new AbortController();
      controller.abort();
      await assert.rejects(
        () => t.call("session.list", {}, { signal: controller.signal }),
        (err: unknown) => {
          assert.ok(err instanceof LoomAbortError);
          assert.strictEqual((err as Error).name, "AbortError");
          return true;
        },
      );
    } finally {
      await t.close();
    }
  });

  test("AbortSignal mid-flight: aborts and emits request.cancel", async () => {
    const cancelCalls: Record<string, unknown>[] = [];
    daemon.registerHandler("request.cancel", (params) => {
      cancelCalls.push(params);
      return { cancelled: true };
    });
    // Slow handler: resolves after 200ms; abort fires at 30ms so the
    // test sees the cancel-driven reject, not the eventual response.
    // .unref() the timer so a leaked one doesn't keep the test runner alive.
    daemon.registerHandler("test.slow", () => {
      return new Promise((resolve) => {
        const t = setTimeout(() => resolve({ done: true }), 200);
        t.unref();
      });
    });

    const t = new LoomTransport(daemon.socketPath, daemon.token);
    await t.connect();
    try {
      const controller = new AbortController();
      const callPromise = t.call("test.slow", {}, { signal: controller.signal });
      // Abort shortly after dispatch.
      setTimeout(() => controller.abort(), 30);
      await assert.rejects(
        () => callPromise,
        (err: unknown) => {
          assert.ok(err instanceof LoomAbortError);
          return true;
        },
      );
      // Give the transport a tick to flush the cancel.
      await new Promise((r) => setTimeout(r, 50));
      assert.ok(cancelCalls.length >= 1, "request.cancel envelope was not sent");
      assert.ok("request_id" in cancelCalls[0]);
    } finally {
      await t.close();
    }
  });
});
